### ISSUE TITLE
Tar reader can't read big layer completely

### ISSUE THREAD

**Issue by Lentil1016** [wagoodman/dive#128]
Hi, thanks for creating this awesome project.

## What happened

I'm trying to run `dive` on a big image(like 2GB). I have sufficient memory on my machine(like 10GB). I found that dive reach the `EOF` before it finish reading some big single layer.

## How to reproduce

```
# I add some output for debuging
$ git diff
diff --git a/image/image.go b/image/image.go
index 3c8baaf..f117fda 100644
--- a/image/image.go
+++ b/image/image.go
@@ -424,6 +424,7 @@ func getFileList(tarredBytes []byte) []filetree.FileInfo {
                header, err := tarReader.Next()
 
                if err == io.EOF {
+                       fmt.Printf("\n\n\n\n\n\n\n\n\n------ EOF size: %d, lenth: %d -------\n\n\n\n\n\n\n\n\n", reader.Size(), reader.Len())
                        break
                }

$ go build  github.com/wagoodman/dive
$ ./dive wnameless/oracle-xe-11g
Analyzing Image
Using default tag: latest
latest: Pulling from wnameless/oracle-xe-11g
Digest: sha256:806c005a3d6af211e92b96015daaeefdcde53b7b6434d34353e620fb46ea7dad
Status: Image is up to date for wnameless/oracle-xe-11g:latest
  Fetching metadata...
  Fetching image... [========================================>] 100 % (2132610935/2132610935)
  Discovering layers... 
------EOF size: 15872, lenth: 0 -------
------EOF size: 8192, lenth: 0 -------
------EOF size: 3072, lenth: 0 -------
------EOF size: 5632, lenth: 0 -------
------EOF size: 275751936, lenth: 0 -------
------EOF size: 82089472, lenth: 0 -------
------EOF size: 1779824640, lenth: 611596800 -------
  Discovering layers... Done!
    ├─ f9e96ffd8b819b8 : [========================================>] 100 % (3/3)
    ├─ ed7958dce50261d : [========================================>] 100 % (333/333)
    ├─ a6cf78075bf5fcb : [========================================>] 100 % (3/3)
    ├─ 6847e59f0ce00e5 : [========================================>] 100 % (3295/3295)
    ├─ 4847692d54ef41d : [========================================>] 100 % (14/14)
    ├─ 29755a695e055b9 : [========================================>] 100 % (10/10)
    ├─ 03c258150a8a7e2 : [========================================>] 100 % (20/20)
  Fetching image config...
  Building tree...
  Analyzing layers...
```

## Additional information

I tried to read file list directly as a file from the same single layer, and this issue dose not come up.

```
package main

import (
	"archive/tar"
	"fmt"
	"io"
	"os"
)

func main() {
	f, err := os.Open("./layer.tar")
	if err != nil {
		fmt.Println(err)
		return
	}
	defer f.Close()
	tarReader := tar.NewReader(f)
	for {
		header, err := tarReader.Next()

		if err == io.EOF {
			break
		}

		if err != nil {
			fmt.Println(err)
			break
		}

		name := header.Name

		switch header.Typeflag {
		case tar.TypeXGlobalHeader:
			fmt.Printf("ERRG: XGlobalHeader: %v: %s\n", header.Typeflag, name)
		case tar.TypeXHeader:
			fmt.Printf("ERRG: XHeader: %v: %s\n", header.Typeflag, name)
		default:
			fmt.Printf("File: %v: %s\n", header.Typeflag, name)
		}
	}
}
```

I'm a go beginner. Is there any more information i can provide? Thanks.

**Comment by wagoodman** [comment]
Thanks, @Lentil1016 ! I'm glad you find dive useful. I won't be able to take a detailed look for a day or so, but the `cache-tree-building` branch solves a few performance issues that you may want to check out ( from #127). Not certain it will solve your problem until I dig further, but it could be worth building on your box and trying it out.

**Comment by Lentil1016** [comment]
That's cool, I'll try it out ASAP.

**Comment by Lentil1016** [comment]
@wagoodman Yes it worked. And even better, switching layer is much more faster. Thanks a lot.

**Comment by wagoodman** [comment]
Good to hear! These fixes (and more) should be in the 0.5.0 release.
